validator/client/runner.go

```diff
@@ -310,39 +310,53 @@ func handleAssignmentError(err error, slot primitives.Slot) {
 	}
 }
 
-func runHealthCheckRoutine(ctx context.Context, v iface.Validator) {
+func runHealthCheckRoutine(ctx context.Context, cancel context.CancelFunc, v iface.Validator, maxHealthChecks int) {
 	log.Info("Starting health check routine for beacon node apis")
-	healthCheckTicker := time.NewTicker(time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second)
+	// just check one a slot
+	interval := time.Duration(params.BeaconConfig().SecondsPerSlot) * time.Second
+	ticker := time.NewTicker(interval)
 	tracker := v.HealthTracker()
 	go func() {
-		// trigger the healthcheck immediately the first time
-		for ; true; <-healthCheckTicker.C {
-			if ctx.Err() != nil {
-				log.WithError(ctx.Err()).Error("Context cancelled")
-				return
-			}
+		defer ticker.Stop()
+
+		// Perform health check and handle host switching/event stream
+		performHealthCheck := func() bool {
 			isHealthy := tracker.CheckHealth(ctx)
 			if !isHealthy && features.Get().EnableBeaconRESTApi {
 				v.ChangeHost()
-				if !tracker.CheckHealth(ctx) {
-					continue // Skip to the next ticker
-				}
-
-				slot, err := v.CanonicalHeadSlot(ctx)
-				if err != nil {
-					log.WithError(err).Error("Could not get canonical head slot")
-					return
-				}
-				if err := v.PushProposerSettings(ctx, slot, true); err != nil {
-					log.WithError(err).Warn("Failed to update proposer settings")
-				}
+				isHealthy = tracker.CheckHealth(ctx) // Re-check after host change
 			}
 
-			// in case of node returning healthy but event stream died
+			// Reconnect event stream if needed
 			if isHealthy && !v.EventStreamIsRunning() {
 				log.Info("Event stream reconnecting...")
 				go v.StartEventStream(ctx, event.DefaultEventTopics)
 			}
+			return isHealthy
+		}
+
+		// Initial check immediately
+		performHealthCheck()
+
+		healthCheckCounter := 0
+		// Continue periodic checks
+		for {
+			select {
+			case <-ticker.C:
+
+				ishealthy := performHealthCheck()
+				if ishealthy {
+					healthCheckCounter = 0
+				} else {
+					healthCheckCounter++
+				}
+				if maxHealthChecks != -1 && healthCheckCounter >= maxHealthChecks {
+					log.Info("Health check timed out")
+					cancel()
```

Comment:
What's the difference in the behavior of the validator client between `cancel` here and `onRunnerExit`?

Reply:
hmm I guess maybe I can switch to that and test it, this is the cancel context on the validator so it would cancel the outside contex and routine ( it's probably safer that way)